File: integration_tests/r_tests/test_analyze.R

# integration tests on doppel-describe commandline entrypoint
# (targeting an R package)
library(jsonlite)
library(testthat)

# details that will always be true of doppel-describe output
TOP_LEVEL_KEYS <- c("name", "language", "functions", "classes")
NUM_TOP_LEVEL_KEYS <- length(TOP_LEVEL_KEYS)

# Set up test data
TEST_PACKAGES <- c(
    'testpkguno'
    , 'testpkgdos'
)
TEST_DATA_DIR <- tools::file_path_as_absolute("../../test_data")

# TL;DR R doesn't care about your bash PATH so finding conda stuff is
# hard. Passing in the direct path to doppel-describe through the environment
DOPPEL_DESCRIBE_LOC <- Sys.getenv("DOPPEL_DESCRIBE_LOC")
if (DOPPEL_DESCRIBE_LOC == ""){
    stop("You did not set environment variable DOPPEL_DESCRIBE_LOC")
}

.run_doppel_describe <- function(package_name, doppel_describe_loc, test_data_dir){

    cmd <- paste0(
        doppel_describe_loc,
        " --language r -p ",
        package_name,
        " --data-dir ",
        test_data_dir
    )

    exit_code <- system(
        command = cmd
        , wait = TRUE
    )

    if (exit_code != 0){
        stop("Running doppel-describe failed")
    }

    OUTPUT_FILE <- paste0("r_", package_name, ".json")
    PATH_TO_OUTPUT_FILE <- file.path(
        test_data_dir
        , OUTPUT_FILE
    )

    # Read in the raw JSON produced by doppel-describe. Using
    # paste0() here to make sure these tests don't care about
    # whether that code pretty-prints the JSON it produces
    raw_json <- paste0(
        readLines(PATH_TO_OUTPUT_FILE)
        , collapse = ""
    )
    result_json <- jsonlite::fromJSON(
        raw_json
        , simplifyVector = TRUE
    )

    return(list(
        "raw" = raw_json
        , "parsed" = result_json
    ))
}

RESULTS <- list()
for (pkg_name in TEST_PACKAGES){
    RESULTS[[pkg_name]] <- .run_doppel_describe(
        package_name = pkg_name
        , doppel_describe_loc = DOPPEL_DESCRIBE_LOC
        , test_data_dir = TEST_DATA_DIR
    )
}

# Small function to take a super long string formatted
# similar to a Python docstring and strip all the whitespace
# generated
.docstring <- function(txt){
    txt <- gsub('\n', '', txt)
    txt <- gsub('\\s+', ' ', txt)
    txt <- trimws(txt)
    return(txt)
}

# Tests that check that basic truths about the
# JSON file produced by doppel-describe remain
# true.
context("Basic contract")

test_that("The JSON file produce by doppel-describe should have only the expected top-level dictionary keys", {

    expect_named(
        RESULTS[["testpkguno"]][["parsed"]]
        , EXPECTED_TOP_LEVEL_KEYS
        , ignore.order = TRUE
        , ignore.case = FALSE
    )
    expect_equal(length(RESULTS[["testpkguno"]][["parsed"]]), NUM_TOP_LEVEL_KEYS)

})

test_that("'name' should be a string", {
    expect_is(RESULTS[["testpkguno"]][["parsed"]][["name"]], "character")
})

test_that("'language' should be 'r'", {
    expect_equal(RESULTS[["testpkguno"]][["parsed"]][["language"]], "r")
})

txt <- .docstring(
    "
    'functions' should be a dictionary key by function name. Each
    function should have a dictionary keyed by 'args' where 'args'
    holds an array of strings.

    Nothing other than 'args' should be included in the function
    interface.
    "
)

test_that(txt, {

    func_block <- RESULTS[["testpkguno"]][["parsed"]][["functions"]]

    func_names <- names(func_block)
    for (func_name in func_names){
        func_interface <- func_block[[func_name]]

        expect_named(
            func_interface
            , "args"
            , ignore.order = TRUE
            , ignore.case = FALSE
        )

        args <- func_interface[["args"]]
        if (length(args) > 0){
            expect_true(is.character(args))
        }
    }

})

txt <- .docstring(
    "
    'classes' should be a dictionary keyed
    by class name. Each of those classes should
    have a single section called 'public_methods'.
    Each method should have a dictionary keyed
    by 'args' where 'args' holds an array of strings.

    Nothing other than 'args' should be included in the
    method interface and nothing other than 'public_methods'
    should be included in the class interface.
    "
)
test_that(txt, {

    class_block <- RESULTS[["testpkguno"]][["parsed"]][['classes']]

    class_names <- names(class_block)
    for (class_name in class_names){
        class_interface <- class_block[[class_name]]

        expect_named(
            class_interface
            , "public_methods"
            , ignore.order = TRUE
            , ignore.case = FALSE
        )

        method_names <- names(class_interface[["public_methods"]])

        for (method_name in method_names){
            method_interface <- class_interface[["public_methods"]][[method_name]]
            args <- method_interface[["args"]]
            if (length(args) > 0){
                expect_true(is.character(args))
            }
        }
    }

})

context("function block")

txt <- .docstring(
    "
    Exported functions should all be found,
    even if decorators are used on them.

    No other stuff should end up in 'functions'.
    "
)
test_that(txt, {

    func_block <- RESULTS[["testpkguno"]][["parsed"]][['functions']]

    expect_named(
        func_block
        , c("function_a", "function_b", "function_c")
        , ignore.order = TRUE
        , ignore.case = FALSE
    )

})

test_that("Functions without arguments should get an 'args' dictionary with an empty list", {
    expect_is(
        RESULTS[["testpkguno"]][["parsed"]][['functions']][['function_a']][['args']]
        , 'list'
    )

    expect_equal(
        RESULTS[["testpkguno"]][["parsed"]][['functions']][['function_a']][['args']]
        , list()
    )
})

test_that("Functions with a mix of actual keyword args and '...' should have correct signature", {
    expect_equal(
        RESULTS[["testpkguno"]][["parsed"]][['functions']][['function_b']]
        , list(
            'args' = c('x', 'y', '~~KWARGS~~')
        )
    )
})

test_that("Functions with only '...' should have the correct signature", {
    expect_equal(
        RESULTS[["testpkguno"]][["parsed"]][['functions']][['function_c']]
        , list(
            'args' = c('~~KWARGS~~')
        )
    )
})

context("class block")

test_that("Exported classes should all be found", {
    expect_named(
        RESULTS[["testpkguno"]][["parsed"]][['classes']]
        , c('ClassA', 'ClassB', 'ClassC', 'ClassD', 'ClassE', 'ClassF')
        , ignore.order = TRUE
        , ignore.case = FALSE
    )
})

txt <- .docstring(
    "
    Public class methods of all exported classes
    should be found.

    No other stuff should end up underneath classes
    within 'classes'.
    "
)
test_that(txt, {

    class_block <- RESULTS[["testpkguno"]][["parsed"]][["classes"]]

    expected_methods <- c(
        '~~CONSTRUCTOR~~',
        'anarchy',
        'banarchy',
        'canarchy'
    )

    expect_named(
        RESULTS[["testpkguno"]][["parsed"]][["classes"]][["ClassA"]][["public_methods"]]
        , expected_methods
        , ignore.order = TRUE
        , ignore.case = FALSE
    )
})


txt <- .docstring(
    "
    Public methods documented in the API of exported
    classes should include methods which are defined
    by a parent object and not overwritten by the
    child.

    No other stuff should end up underneath classes
    within 'classes'.
    "
)
test_that(txt, {

    class_block <- RESULTS[["testpkguno"]][["parsed"]][["classes"]]

    expected_methods <- c(
        '~~CONSTRUCTOR~~',
        'anarchy',
        'banarchy',
        'canarchy',
        'hello_there'
    )

    expect_named(
        RESULTS[["testpkguno"]][["parsed"]][["classes"]][["ClassB"]][["public_methods"]]
        , expected_methods
        , ignore.order = TRUE
        , ignore.case = FALSE
    )

})

txt <- .docstring(
    "
    Class methods should be correctly found and
    documented alongside other public methods in
    a class
    "
)
test_that(txt, {
    expect_named(
        RESULTS[["testpkguno"]][["parsed"]][["classes"]][["ClassC"]][["public_methods"]]
        , c("~~CONSTRUCTOR~~", "from_string")
        , ignore.order = TRUE
        , ignore.case = FALSE
    )
})

txt <- .docstring(
    "
    Class methods inherited from a parent class should be correctly found
    and documented alongside other public methods in a class
    "
)
test_that(txt, {
    testthat::skip("This test does not apply to R6 classes. You cannot inherit class methods in R6.")
})

test_that("analyze.R should ignore R6-specific public methods like print()", {
    expect_named(
        RESULTS[["testpkguno"]][["parsed"]][["classes"]][["ClassD"]][["public_methods"]]
        , c("~~CONSTRUCTOR~~", "from_string")
        , ignore.order = TRUE
        , ignore.case = FALSE
    )
})

test_that("Classes with constructors that have no keyword args should be serialized correctly", {
    expect_named(
        RESULTS[["testpkguno"]][["parsed"]][["classes"]][["ClassE"]][["public_methods"]]
        , c("~~CONSTRUCTOR~~", "from_string")
        , ignore.order = TRUE
        , ignore.case = FALSE
    )
    expect_is(RESULTS[["testpkguno"]][["parsed"]][["classes"]][["ClassE"]][["public_methods"]][["~~CONSTRUCTOR~~"]][["args"]], "list")
    expect_is(RESULTS[["testpkguno"]][["parsed"]][["classes"]][["ClassE"]][["public_methods"]][["from_string"]][["args"]], "list")

    # test that things with no kwargs produce "args": [], not "args": {}
    expect_true(isTRUE(
        grepl('.+"ClassE".+~~CONSTRUCTOR~~.+"args"\\:\\[\\]', RESULTS[["testpkguno"]][["raw"]])
    ))
    expect_true(isTRUE(
        grepl('.+"from_string".+~~CONSTRUCTOR~~.+"args"\\:\\[\\]', RESULTS[["testpkguno"]][["raw"]])
    ))
})

test_that("Totally empty classes should still have their constructors documented", {
    expect_named(
        RESULTS[["testpkguno"]][["parsed"]][["classes"]][["ClassF"]][["public_methods"]]
        , c("~~CONSTRUCTOR~~")
        , ignore.order = TRUE
        , ignore.case = FALSE
    )
    expect_named(
        RESULTS[["testpkguno"]][["parsed"]][["classes"]][["ClassF"]][["public_methods"]][["~~CONSTRUCTOR~~"]]
        , c("args")
        , ignore.order = TRUE
        , ignore.case = FALSE
    )
})

context("function-only packages")

test_that("The JSON file produce by doppel-describe should have only the expected top-level dictionary keys", {

    expect_named(
        RESULTS[["testpkgdos"]][["parsed"]]
        , EXPECTED_TOP_LEVEL_KEYS
        , ignore.order = TRUE
        , ignore.case = FALSE
    )
    expect_equal(length(RESULTS[["testpkguno"]][["parsed"]]), NUM_TOP_LEVEL_KEYS)

})

File: integration_tests/python_tests/test_analyze.py
"""
integration tests on doppel-describe
commandline entrypoint (targeting a
python package)
"""

import json
import os
import pytest

# details that will always be true of doppel-describe output
EXPECTED_TOP_LEVEL_KEYS = set([
    "name",
    "language",
    "functions",
    "classes"
])
NUM_TOP_LEVEL_KEYS = len(EXPECTED_TOP_LEVEL_KEYS)

with open('../../doppel/VERSION', 'r') as f:
    EXPECTED_VERSION = f.read().strip()


@pytest.fixture()
def rundescribe():
    """
    run doppel-describe to generate output file
    """
    # there isn't a clean way to pass in
    # command-line args to test scripts, so
    # using environment variables
    test_packages = [
        'testpkguno',
        'testpkgdos',
        'testpkgtres',
        'pythonspecific',
        'pythonspecific2'
    ]

    # Added this abomination because something about
    # os.getenv('TEST_PACKAGE_DIR') was resulting in a None
    test_data_dir = os.path.abspath('../../test_data')

    results = {}

    for package_name in test_packages:
        cmd = "doppel-describe --language python -p {} --data-dir {}".format(
            package_name,
            test_data_dir
        )

        exit_code = os.system(cmd)
        if exit_code != 0:
            msg = "doppel-describe exited with non-zero exit code: {}"
            raise RuntimeError(msg.format(exit_code))

        output_file = "python_{}.json".format(package_name)
        path_to_output_file = os.path.join(
            test_data_dir,
            output_file
        )
        with open(path_to_output_file, 'r') as f:
            result_json = json.loads(f.read())

        results[package_name] = result_json

    return results


class TestVersion:
    """
    --version should work as expected.
    """

    def test_version(self):
        """
        doppel-test --version should work
        """
        version_string = os.popen('doppel-test --version').read().strip()
        assert version_string == EXPECTED_VERSION

    def test_describe_version(self):
        """
        doppel-describe --version should work
        """
        version_string = os.popen('doppel-describe --version').read().strip()
        assert version_string == EXPECTED_VERSION


class TestBasicContract:
    """
    Tests that check that basic truths about the
    JSON file produced by doppel-describe remain
    true.
    """

    def test_contract(self, rundescribe):
        """
        The JSON file produced by doppel-describe should have
        only the expected top-level dictionary keys
        """
        result_json = rundescribe['testpkguno']

        for top_level_key in EXPECTED_TOP_LEVEL_KEYS:
            assert result_json.get(top_level_key, False)
        assert len(result_json.keys()) == NUM_TOP_LEVEL_KEYS

    def test_name(self, rundescribe):
        """
        'name' should be a string
        """
        assert isinstance(rundescribe['testpkguno']['name'], str)

    def test_language(self, rundescribe):
        """
        'language' should be 'python'
        """
        assert rundescribe['testpkguno']['language'] == 'python'

    def test_functions_block(self, rundescribe):
        """
        'functions' should be a dictionary keyed
        by function name. Each function should have a dictionary keyed
        by 'args' where 'args' holds an array of strings.

        Nothing other than 'args' should be included in the
        function interface.
        """
        for func_name, func_interface in rundescribe['testpkguno']['functions'].items():
            args = func_interface['args']
            assert isinstance(args, list)
            assert len(func_interface.keys()) == 1
            if len(args) > 0:
                assert all([isinstance(x, str) for x in args])

    def test_classes_block(self, rundescribe):
        """
        'classes' should be a dictionary keyed
        by class name. Each of those classes should
        have a single section called 'public_methods'.
        Each method should have a dictionary keyed
        by 'args' where 'args' holds an array of strings.

        Nothing other than 'args' should be included in the
        method interface and nothing other than 'public_methods'
        should be included in the class interface.
        """
        for class_name, class_interface in rundescribe['testpkguno']['classes'].items():

            assert len(class_interface.keys()) == 1

            for method_name, method_interface in class_interface['public_methods'].items():
                args = method_interface['args']
                assert isinstance(args, list)
                assert len(method_interface.keys()) == 1
                if len(args) > 0:
                    assert all([isinstance(x, str) for x in args])


class TestFunctionStuff:
    """
    Tests that the "function" block of the JSON
    produced by doppel-describe is correct.
    """

    def test_functions_found(self, rundescribe):
        """
        Exported functions should all be found,
        even if decorators are used on them.

        No other stuff should end up in "functions".
        """
        func_dict = rundescribe['testpkguno']['functions']

        expected_functions = [
            'function_a',
            'function_b',
            'function_c'
        ]

        for f in expected_functions:
            assert func_dict.get(f, False)

        assert len(func_dict.keys()) == len(expected_functions)

    def test_empty_function(self, rundescribe):
        """
        Functions without any arguments should get an
        'args' dictionary with an empty list.
        """
        assert rundescribe['testpkguno']['functions']['function_a']['args'] == []

    def test_regular_function(self, rundescribe):
        """
        Functions with a mix of actual keyword args
        and '**kwargs' should have the correct signature.
        """
        expected = {
            "args": ['x', 'y', '~~KWARGS~~']
        }
        assert rundescribe['testpkguno']['functions']['function_b'] == expected

    def test_kwargs_only_function(self, rundescribe):
        """
        Functions with only '**kwargs' should have
        the correct signature.
        """
        expected = {
            "args": ['~~KWARGS~~']
        }
        assert rundescribe['testpkguno']['functions']['function_c'] == expected


class TestClassStuff:
    """
    Tests that the "classes" block of the JSON
    produced by doppel-describe is correct.
    """

    def test_classes_found(self, rundescribe):
        """
        Exported classes should all be found.
        """
        class_dict = rundescribe['testpkguno']['classes']

        expected_classes = [
            'ClassA',
            'ClassB',
            'ClassC',
            'ClassD',
            'ClassE',
            'ClassF'
        ]

        for c in expected_classes:
            assert class_dict.get(c, False)

        assert len(class_dict.keys()) == len(expected_classes)

    def test_class_public_methods_found(self, rundescribe):
        """
        Public class methods of all exported classes
        should be found.

        No other stuff should end up underneath classes
        within "classes".
        """
        class_dict = rundescribe['testpkguno']['classes']
        expected_methods = [
            '~~CONSTRUCTOR~~',
            'anarchy',
            'banarchy',
            'canarchy'
        ]

        for e in expected_methods:
            assert class_dict['ClassA']['public_methods'].get(e, False)

        assert len(class_dict['ClassA']['public_methods'].keys()) == len(expected_methods)

    def test_inherited_class_public_methods_found(self, rundescribe):
        """
        Public methods documented in the API of exported
        classes should include methods which are defined
        by a parent object and not overwritten by the
        child.

        No other stuff should end up underneath classes
        within "classes".
        """
        class_dict = rundescribe['testpkguno']['classes']
        expected_methods = [
            '~~CONSTRUCTOR~~',
            'anarchy',
            'banarchy',
            'canarchy',
            'hello_there'
        ]

        for e in expected_methods:
            assert class_dict['ClassB']['public_methods'].get(e, False)

        assert len(class_dict['ClassB']['public_methods'].keys()) == len(expected_methods)

    def test_classmethods_found(self, rundescribe):
        """
        Class methods should be correctly found and
        documented alongside other public methods in
        a class
        """
        assert rundescribe['testpkguno']['classes']['ClassC']['public_methods'].get('from_string', False)

    def test_inherited_classmethods_found(self, rundescribe):
        """
        Class methods inherited from a parent class
        should be correctly found and documented
        alongside other public methods in a class
        """
        assert rundescribe['testpkguno']['classes']['ClassD']['public_methods'].get('from_string', False)

    def test_empty_constructors(self, rundescribe):
        """
        Classes with constructors that have no keyword args
        should be serialized correctly
        """
        class_dict = rundescribe['testpkguno']['classes']
        expected_methods = [
            '~~CONSTRUCTOR~~',
            'from_string'
        ]

        for e in expected_methods:
            assert class_dict['ClassE']['public_methods'].get(e, False)

        # test that things with no kwargs produce "args": [], not "args": {}
        # expect_true(isTRUE(
        #    grepl('.+"ClassE".+~~CONSTRUCTOR~~.+"args"\\:\\[\\]', RESULTS[["testpkguno"]][["raw"]])
        # ))
        # expect_true(isTRUE(
        #     grepl('.+"from_string".+~~CONSTRUCTOR~~.+"args"\\:\\[\\]', RESULTS[["testpkguno"]][["raw"]])
        # ))

    def test_empty_classes(self, rundescribe):
        """
        Totally empty classes should still have their
        constructors documented
        """
        assert list(rundescribe['testpkguno']['classes']['ClassF']['public_methods'].keys()) == ['~~CONSTRUCTOR~~']
        assert rundescribe['testpkguno']['classes']['ClassF']['public_methods']['~~CONSTRUCTOR~~'] == {'args': []}


class TestFunctionOnly:
    """
    Test the behavior of analyze.py for packages
    which have functions but not classes
    """

    def test_top_level_keys(self, rundescribe):
        """
        The JSON file produce by doppel-describe
        should have only the expected top-level dictionary keys
        """
        result_json = rundescribe['testpkgdos']

        for top_level_key in EXPECTED_TOP_LEVEL_KEYS:
            assert result_json.get(top_level_key, None) is not None
        assert len(result_json.keys()) == NUM_TOP_LEVEL_KEYS


class TestClassOnly:
    """
    Test the behavior of analyze.py for packages
    which have classes but not functions
    """

    def test_top_level_keys(self, rundescribe):
        """
        The JSON file produce by doppel-describe
        should have only the expected top-level dictionary keys
        """
        result_json = rundescribe['testpkgtres']

        for top_level_key in EXPECTED_TOP_LEVEL_KEYS:
            assert result_json.get(top_level_key, None) is not None
        assert len(result_json.keys()) == NUM_TOP_LEVEL_KEYS


class TestPythonSpecific:
    """
    Test the behavior of analyze.py for packages
    with some Python-specific features like
    submodules and custom exceptions
    """

    def test_top_level_keys(self, rundescribe):
        """
        The JSON file produce by doppel-describe
        should have only the expected top-level dictionary keys
        """
        result_json = rundescribe['pythonspecific']

        for top_level_key in EXPECTED_TOP_LEVEL_KEYS:
            assert result_json.get(top_level_key, None) is not None
        assert len(result_json.keys()) == NUM_TOP_LEVEL_KEYS

    def test_sub_modules(self, rundescribe):
        """
        analyze.py should correctly handle python submodules and
        should ignore package constant.
        """
        result_json = rundescribe['pythonspecific']

        assert set(result_json['functions'].keys()) == set(['some_function', 'wrap_min'])
        assert set(result_json['classes'].keys()) == set(['SomeClass', 'GreatClass', 'MinWrapper'])

    def test_inner_classes(self, rundescribe):
        """
        analyze.py should correctly handle classes
        that are included as members of another
        class
        """
        result_json = rundescribe['pythonspecific']

        assert set(result_json['classes']['GreatClass']['public_methods'].keys()) == set(['do_stuff', 'LilGreatClass', '~~CONSTRUCTOR~~'])
        lil_args = result_json['classes']['GreatClass']['public_methods']['LilGreatClass']['args']
        assert set(lil_args) == set(['things', 'stuff'])

    def test_builtin_func(self, rundescribe):
        """
        analyze.py should correctly handle the case where a built-in
        like min() has been mapped directly to an exported function
        """
        result_json = rundescribe['pythonspecific']

        assert result_json['functions']['wrap_min'] == {'args': []}

    def test_builtin_method(self, rundescribe):
        """
        analyze.py should correctly handle the case where a built-in
        like min() has been mapped directly to a public method
        of a class
        """
        result_json = rundescribe['pythonspecific']

        assert result_json['classes']['MinWrapper']['public_methods']['wrap_min'] == {'args': []}


class TestPythonSpecific:
    """
    Test the behavior of analyze.py for packages using
    'from <module> import *' in __init__.py. This package also
    tests the behavior of analyze.py in the presence of top-level
    built-in imports like "from warnings import warn"
    """

    def test_top_level_keys(self, rundescribe):
        """
        The JSON file produce by doppel-describe
        should have only the expected top-level dictionary keys
        """
        result_json = rundescribe['pythonspecific2']

        for top_level_key in EXPECTED_TOP_LEVEL_KEYS:
            assert result_json.get(top_level_key, None) is not None
        assert len(result_json.keys()) == NUM_TOP_LEVEL_KEYS

    def test_functions(self, rundescribe):
        """
        analyze.py should find all functions defined in the package
        whose names do not start with "_", regardless of what is in
        ``__all__`` in ``__init__.py``.
        """
        result_json = rundescribe['pythonspecific2']

        assert set(result_json['functions'].keys()) == set(['create_warning', 'create_warm_things'])
        # imports from other packages are included if you explicitly
        # wrap them in a def()
        assert 'create_warm_things' in result_json['functions']
        # import from other packages are excluded even if you map them to
        # a new name in your package
        assert 'shmeate_schmarning' not in result_json['functions']
        # internal functions
        assert '_super_secret' not in result_json['functions'].keys()
        # imported standard lib function
        assert 'warn' not in result_json['functions'].keys()
        # imported non-standard-lib function
        assert 'get' not in result_json['functions'].keys()
        # imports from other packages are ignored even if you rename them
        # or add them to __all__ in __init__.py
        assert 'custom_post' not in result_json['functions'].keys()
        assert 'post' not in result_json['functions'].keys()

    def test_classes(self, rundescribe):
        """
        analyze.py should not have found any classes in this
        package but should have created the 'classes' section
        """
        result_json = rundescribe['pythonspecific2']
        assert result_json['classes'] == {}
